// ircc-ds-angular-component-library/esm2020/lib/banner-component/banner/banner.service.mjs
import { Injectable } from '@angular/core';
import { BehaviorSubject } from 'rxjs';
import * as i0 from "@angular/core";
export class BannerService {
    constructor() {
        this.bannerSubj = new BehaviorSubject({
            id: ''
        });
        this.bannerSubjObs$ = this.bannerSubj.asObservable();
        this.toggleSubj = new BehaviorSubject({
            id: '',
            value: false
        });
        this.toggleSubjObs$ = this.toggleSubj.asObservable();
        this.setBanner = (update) => {
            this.bannerSubj.next(update);
        };
        this.toggleBanner = (id, value) => {
            this.toggleSubj.next({ id, value });
        };
    }
}
BannerService.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.2.12", ngImport: i0, type: BannerService, deps: [], target: i0.ɵɵFactoryTarget.Injectable });
BannerService.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "14.2.12", ngImport: i0, type: BannerService, providedIn: 'root' });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.2.12", ngImport: i0, type: BannerService, decorators: [{
            type: Injectable,
            args: [{
                    providedIn: 'root'
                }]
        }] });
