// ircc-ds-angular-component-library/esm2020/lib/shared/label/label.component.mjs
import { Component, Input } from '@angular/core';
import { FormGroup } from '@angular/forms';
import * as i0 from "@angular/core";
import * as i1 from "@ngx-translate/core";
import * as i2 from "../../../shared/functions/stand-alone.functions";
import * as i3 from "./label-button.service";
import * as i4 from "@angular/common";
export const ERROR_TEXT_STUB = {
    en: 'Error',
    fr: 'Erreur'
};
export const HELP_ICON_ALT = {
    en: ', more information',
    fr: ", plus d'information"
};
export class LabelComponent {
    constructor(translate, standAloneFunctions, labelButton) {
        this.translate = translate;
        this.standAloneFunctions = standAloneFunctions;
        this.labelButton = labelButton;
        this.config = {
            formGroup: new FormGroup({}),
            parentID: '',
            legend: false
        };
        this.labelIconText = '';
    }
    ngOnInit() {
        //set config from individual options, if present
        if (this.formGroup)
            this.config.formGroup = this.formGroup;
        if (this.errorMessages)
            this.config.errorMessages = this.errorMessages;
        if (this.parentID)
            this.config.parentID = this.parentID;
        if (this.label)
            this.config.label = this.label;
        if (this.desc)
            this.config.desc = this.desc;
        if (this.hint)
            this.config.hint = this.hint;
        if (this.required)
            this.config.required = this.required;
        if (this.iconButton)
            this.config.iconButton = this.iconButton;
        if (this.topLabel)
            this.config.topLabel = this.topLabel;
        if (this.touched)
            this.config.touched = this.touched;
        if (this.legend)
            this.config.legend = this.legend;
        this.setLang(this.translate.currentLang);
        this.translate.onLangChange.subscribe((change) => {
            this.setLang(change.lang);
        });
    }
    setLang(lang) {
        if (lang === 'en' || lang === 'en-US') {
            // this.errorStubText = ERROR_TEXT_STUB_EN;
            this.labelIconText = HELP_ICON_ALT.en;
        }
        else {
            // this.errorStubText = ERROR_TEXT_STUB_FR;
            this.labelIconText = HELP_ICON_ALT.fr;
        }
    }
    /**
     * Output the button press
     * @param id of the button being pressed (same as component ID)
     */
    iconButtonClick() {
        this.labelButton.buttonPress(this.config.parentID);
    }
    returnLabel() {
        return !this.config.topLabel ? this.config.label : this.config.topLabel;
    }
}
LabelComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.2.12", ngImport: i0, type: LabelComponent, deps: [{ token: i1.TranslateService }, { token: i2.StandAloneFunctions }, { token: i3.LabelButtonService }], target: i0.ɵɵFactoryTarget.Component });
LabelComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "14.2.12", type: LabelComponent, selector: "ircc-cl-lib-label", inputs: { config: "config", formGroup: "formGroup", errorMessages: "errorMessages", parentID: "parentID", label: "label", desc: "desc", hint: "hint", required: "required", iconButton: "iconButton", topLabel: "topLabel", touched: "touched", legend: "legend" }, ngImport: i0, template: "<div\n  id=\"{{ config.parentID + '_label' }}\"\n  class=\"label_container\"\n  [ngClass]=\"{ extra_padding: config.label && !config.desc && !config.hint }\"\n>\n  <ng-container *ngIf=\"config.legend === true; else noLegend\">\n    <ng-container *ngIf=\"config.label\">\n      <div\n        *ngIf=\"config.label\"\n        class=\"label\"\n        [attr.for]=\"config.parentID\"\n      >\n        <div class=\"label_field_container\">\n          <div\n            class=\"required-field-container\"\n            *ngIf=\"config.required\"\n          >\n            <i class=\"fa-regular fa-asterisk required-star\"></i>\n            {{ config.label || '' | translate }}\n          </div>\n          <div *ngIf=\"!config.required\">\n            {{ config.label || '' | translate }}\n          </div>\n          <div\n            *ngIf=\"config.iconButton\"\n            class=\"icon_container\"\n          >\n            <button\n              class=\"touch_button\"\n              category=\"plain\"\n              (click)=\"iconButtonClick()\"\n              role=\"button\"\n              [attr.aria-label]=\"\n                (config.label | translate) +\n                ' ' +\n                (config.iconButton.ariaText | translate)\n              \"\n            ></button>\n            <i [class]=\"config.iconButton.iconClass\"></i>\n          </div>\n        </div>\n      </div>\n    </ng-container>\n  </ng-container>\n  <p\n    *ngIf=\"config.desc\"\n    class=\"input-desc\"\n  >\n    {{ config.desc || '' | translate }}\n  </p>\n  <p\n    *ngIf=\"config.hint\"\n    class=\"input-hint\"\n  >\n    {{ config.hint | translate }}\n  </p>\n</div>\n\n<ng-template #noLegend>\n  <label\n    *ngIf=\"config.label\"\n    class=\"label\"\n    [attr.for]=\"config.parentID\"\n  >\n    <span class=\"label_field_container\">\n      <div\n        class=\"required-field-container\"\n        *ngIf=\"config.required\"\n      >\n        <i class=\"fa-regular fa-asterisk required-star\"></i\n        >{{ config.label || '' | translate }}\n      </div>\n      <div *ngIf=\"!config.required\">\n        {{ config.label || '' | translate }}\n      </div>\n      <div\n        *ngIf=\"config.iconButton\"\n        class=\"icon_container\"\n      >\n        <button\n          class=\"touch_button\"\n          category=\"plain\"\n          (click)=\"iconButtonClick()\"\n          role=\"button\"\n          [attr.aria-label]=\"\n            (config.label | translate) +\n            ' ' +\n            (config.iconButton.ariaText | translate)\n          \"\n        ></button>\n        <i [class]=\"config.iconButton.iconClass\"></i>\n      </div>\n    </span>\n  </label>\n</ng-template>\n", dependencies: [{ kind: "directive", type: i4.NgClass, selector: "[ngClass]", inputs: ["class", "ngClass"] }, { kind: "directive", type: i4.NgIf, selector: "[ngIf]", inputs: ["ngIf", "ngIfThen", "ngIfElse"] }, { kind: "pipe", type: i1.TranslatePipe, name: "translate" }], preserveWhitespaces: true });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.2.12", ngImport: i0, type: LabelComponent, decorators: [{
            type: Component,
            args: [{ selector: 'ircc-cl-lib-label', template: "<div\n  id=\"{{ config.parentID + '_label' }}\"\n  class=\"label_container\"\n  [ngClass]=\"{ extra_padding: config.label && !config.desc && !config.hint }\"\n>\n  <ng-container *ngIf=\"config.legend === true; else noLegend\">\n    <ng-container *ngIf=\"config.label\">\n      <div\n        *ngIf=\"config.label\"\n        class=\"label\"\n        [attr.for]=\"config.parentID\"\n      >\n        <div class=\"label_field_container\">\n          <div\n            class=\"required-field-container\"\n            *ngIf=\"config.required\"\n          >\n            <i class=\"fa-regular fa-asterisk required-star\"></i>\n            {{ config.label || '' | translate }}\n          </div>\n          <div *ngIf=\"!config.required\">\n            {{ config.label || '' | translate }}\n          </div>\n          <div\n            *ngIf=\"config.iconButton\"\n            class=\"icon_container\"\n          >\n            <button\n              class=\"touch_button\"\n              category=\"plain\"\n              (click)=\"iconButtonClick()\"\n              role=\"button\"\n              [attr.aria-label]=\"\n                (config.label | translate) +\n                ' ' +\n                (config.iconButton.ariaText | translate)\n              \"\n            ></button>\n            <i [class]=\"config.iconButton.iconClass\"></i>\n          </div>\n        </div>\n      </div>\n    </ng-container>\n  </ng-container>\n  <p\n    *ngIf=\"config.desc\"\n    class=\"input-desc\"\n  >\n    {{ config.desc || '' | translate }}\n  </p>\n  <p\n    *ngIf=\"config.hint\"\n    class=\"input-hint\"\n  >\n    {{ config.hint | translate }}\n  </p>\n</div>\n\n<ng-template #noLegend>\n  <label\n    *ngIf=\"config.label\"\n    class=\"label\"\n    [attr.for]=\"config.parentID\"\n  >\n    <span class=\"label_field_container\">\n      <div\n        class=\"required-field-container\"\n        *ngIf=\"config.required\"\n      >\n        <i class=\"fa-regular fa-asterisk required-star\"></i\n        >{{ config.label || '' | translate }}\n      </div>\n      <div *ngIf=\"!config.required\">\n        {{ config.label || '' | translate }}\n      </div>\n      <div\n        *ngIf=\"config.iconButton\"\n        class=\"icon_container\"\n      >\n        <button\n          class=\"touch_button\"\n          category=\"plain\"\n          (click)=\"iconButtonClick()\"\n          role=\"button\"\n          [attr.aria-label]=\"\n            (config.label | translate) +\n            ' ' +\n            (config.iconButton.ariaText | translate)\n          \"\n        ></button>\n        <i [class]=\"config.iconButton.iconClass\"></i>\n      </div>\n    </span>\n  </label>\n</ng-template>\n" }]
        }], ctorParameters: function () { return [{ type: i1.TranslateService }, { type: i2.StandAloneFunctions }, { type: i3.LabelButtonService }]; }, propDecorators: { config: [{
                type: Input
            }], formGroup: [{
                type: Input
            }], errorMessages: [{
                type: Input
            }], parentID: [{
                type: Input
            }], label: [{
                type: Input
            }], desc: [{
                type: Input
            }], hint: [{
                type: Input
            }], required: [{
                type: Input
            }], iconButton: [{
                type: Input
            }], topLabel: [{
                type: Input
            }], touched: [{
                type: Input
            }], legend: [{
                type: Input
            }] } });
