// ircc-ds-angular-component-library/esm2020/shared/pipes/autocomplete.pipe.mjs
import { Pipe, SecurityContext } from '@angular/core';
import * as i0 from "@angular/core";
import * as i1 from "@angular/platform-browser";
export class autocompletePipe {
    constructor(domSanitizer) {
        this.domSanitizer = domSanitizer;
    }
    transform(value, searched = '') {
        if (searched.length === 0) {
            return this.sanitize(`${value}`);
        }
        return this.sanitize(`<span class="bolded">${this.replace(value, searched)}</span>`);
    }
    replace(str, searched) {
        return str.replace(new RegExp(`(${searched})`, 'i'), '<span class="normal">$1</span>');
    }
    sanitize(str) {
        return this.domSanitizer.sanitize(SecurityContext.HTML, str);
    }
}
autocompletePipe.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.2.12", ngImport: i0, type: autocompletePipe, deps: [{ token: i1.DomSanitizer }], target: i0.ɵɵFactoryTarget.Pipe });
autocompletePipe.ɵpipe = i0.ɵɵngDeclarePipe({ minVersion: "14.0.0", version: "14.2.12", ngImport: i0, type: autocompletePipe, name: "autocomplete" });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.2.12", ngImport: i0, type: autocompletePipe, decorators: [{
            type: Pipe,
            args: [{ name: 'autocomplete' }]
        }], ctorParameters: function () { return [{ type: i1.DomSanitizer }]; } });
