// ircc-ds-angular-component-library/esm2020/shared/pipes/pipes.module.mjs
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { TranslateModule } from '@ngx-translate/core';
import { RouterModule } from '@angular/router';
import { autocompletePipe } from './autocomplete.pipe';
import * as i0 from "@angular/core";
const IrccDsSharedComponents = [autocompletePipe];
export class IrccDsAngularPipesdModule {
}
IrccDsAngularPipesdModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.2.12", ngImport: i0, type: IrccDsAngularPipesdModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
IrccDsAngularPipesdModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "14.2.12", ngImport: i0, type: IrccDsAngularPipesdModule, declarations: [autocompletePipe], imports: [CommonModule, TranslateModule, RouterModule], exports: [CommonModule,
        TranslateModule,
        RouterModule, autocompletePipe] });
IrccDsAngularPipesdModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "14.2.12", ngImport: i0, type: IrccDsAngularPipesdModule, imports: [CommonModule, TranslateModule, RouterModule, CommonModule,
        TranslateModule,
        RouterModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.2.12", ngImport: i0, type: IrccDsAngularPipesdModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [...IrccDsSharedComponents],
                    imports: [CommonModule, TranslateModule, RouterModule],
                    exports: [
                        CommonModule,
                        TranslateModule,
                        RouterModule,
                        ...IrccDsSharedComponents
                    ]
                }]
        }] });
