// ircc-ds-angular-component-library/esm2020/lib/header-footer/theme-switch/theme-switch.component.mjs
import { Component } from '@angular/core';
import * as i0 from "@angular/core";
// import { ThemeSwitchService } from './theme-switch.service';
export class ThemeSwitchComponent {
    constructor(renderer) {
        this.renderer = renderer;
        this.isDarkTheme = false;
        this.currentTheme = '';
        this.previousTheme = '';
    }
    ngOnInit() {
        this.darkModeQuery = window.matchMedia('(prefers-color-scheme: dark)');
        this.isDarkTheme = this.darkModeQuery.matches;
        this.isDarkTheme ? this.currentTheme = 'dark' : this.currentTheme = 'light';
        this.currentTheme === 'dark' ? this.previousTheme = 'Light' : this.previousTheme = 'Dark';
        this.darkModeQuery.addEventListener('change', this.handleDarkModeChange.bind(this));
    }
    toggleTheme() {
        this.isDarkTheme = !this.isDarkTheme;
        this.isDarkTheme ? this.currentTheme = 'dark' : this.currentTheme = 'light';
        this.currentTheme === 'dark' ? this.previousTheme = 'Light' : this.previousTheme = 'Dark';
        this.renderer.setAttribute(document.documentElement, 'scheme', this.currentTheme);
        this.darkModeQuery.removeEventListener('change', this.handleDarkModeChange.bind(this));
    }
    handleDarkModeChange(event) {
        this.isDarkTheme = event.matches;
        this.isDarkTheme ? this.currentTheme = 'dark' : this.currentTheme = 'light';
        this.currentTheme === 'dark' ? this.previousTheme = 'Light' : this.previousTheme = 'Dark';
    }
}
ThemeSwitchComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.2.12", ngImport: i0, type: ThemeSwitchComponent, deps: [{ token: i0.Renderer2 }], target: i0.ɵɵFactoryTarget.Component });
ThemeSwitchComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "14.2.12", type: ThemeSwitchComponent, selector: "ircc-cl-lib-theme-switch", ngImport: i0, template: "<button\n  (click)=\"toggleTheme()\"\n  category=\"secondary\"\n  color=\"critical\"\n>\n  {{ this.previousTheme }}\n</button>\n", preserveWhitespaces: true });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.2.12", ngImport: i0, type: ThemeSwitchComponent, decorators: [{
            type: Component,
            args: [{ selector: 'ircc-cl-lib-theme-switch', template: "<button\n  (click)=\"toggleTheme()\"\n  category=\"secondary\"\n  color=\"critical\"\n>\n  {{ this.previousTheme }}\n</button>\n" }]
        }], ctorParameters: function () { return [{ type: i0.Renderer2 }]; } });
