// ircc-ds-angular-component-library/esm2020/shared/constants/jl-components.constants.mjs
export var DropdownTypes;
(function (DropdownTypes) {
    DropdownTypes["input"] = "input";
    DropdownTypes["cta"] = "cta";
})(DropdownTypes || (DropdownTypes = {}));
export var DropDownActions;
(function (DropDownActions) {
    DropDownActions["addApplications"] = "add-applications";
    DropDownActions["filterApplications"] = "filter-applications";
})(DropDownActions || (DropDownActions = {}));
export var DSSizes;
(function (DSSizes) {
    DSSizes["large"] = "large";
    DSSizes["small"] = "small";
})(DSSizes || (DSSizes = {}));
export var DSFullSizes;
(function (DSFullSizes) {
    DSFullSizes["large"] = "large";
    DSFullSizes["small"] = "small";
    DSFullSizes["extraSmall"] = "extraSmall";
})(DSFullSizes || (DSFullSizes = {}));
export var DSOrientations;
(function (DSOrientations) {
    DSOrientations["horizontal"] = "horizontal";
    DSOrientations["vertical"] = "vertical";
})(DSOrientations || (DSOrientations = {}));
export var DSViewPortSize;
(function (DSViewPortSize) {
    DSViewPortSize[DSViewPortSize["default"] = 0] = "default";
    DSViewPortSize[DSViewPortSize["mobile"] = 768] = "mobile";
    DSViewPortSize[DSViewPortSize["tablet"] = 992] = "tablet";
    DSViewPortSize[DSViewPortSize["desktop"] = 1366] = "desktop";
})(DSViewPortSize || (DSViewPortSize = {}));
