// ircc-ds-angular-component-library/esm2020/lib/header-footer/footer/footer.component.mjs
import { Component, Input } from '@angular/core';
import { GOV_CANADA_LOGOS } from '../header-footer-const.component';
import * as i0 from "@angular/core";
import * as i1 from "@ngx-translate/core";
import * as i2 from "../language-switch/language-header-footer-switch.service";
export const GOV_LOGO_ALT_TEXT_EN = 'Canada wordmark';
export const GOV_LOGO_ALT_TEXT_FR = 'FR Canada wordmark';
export class FooterComponent {
    constructor(translate, languageHeaderFooterSwitch) {
        this.translate = translate;
        this.languageHeaderFooterSwitch = languageHeaderFooterSwitch;
        this.id = '';
        this.altImage = '';
        this.logo = GOV_CANADA_LOGOS.footerLightLogo;
        this.isDarkMode = false;
        this.subscription = this.languageHeaderFooterSwitch.isDarkMode$.subscribe((response) => {
            this.updateFooterImage(response);
            this.isDarkMode = response;
        });
    }
    ngOnInit() {
        this.setLang(this.translate.currentLang);
        this.translate.onLangChange.subscribe((change) => {
            this.setLang(change.lang);
            this.updateFooterImage(this.isDarkMode);
        });
    }
    updateFooterImage(res) {
        this.logo = res
            ? GOV_CANADA_LOGOS.footerDarkLogo
            : GOV_CANADA_LOGOS.footerLightLogo;
    }
    setLang(lang) {
        if (lang === 'en' || lang === 'en-US') {
            this.altImage = GOV_LOGO_ALT_TEXT_EN;
        }
        else {
            this.altImage = GOV_LOGO_ALT_TEXT_FR;
        }
    }
    ngOnDestroy() {
        this.subscription.unsubscribe();
    }
}
FooterComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.2.12", ngImport: i0, type: FooterComponent, deps: [{ token: i1.TranslateService }, { token: i2.LanguageHeaderFooterSwitchService }], target: i0.ɵɵFactoryTarget.Component });
FooterComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "14.2.12", type: FooterComponent, selector: "ircc-cl-lib-footer", inputs: { id: "id" }, ngImport: i0, template: "<footer\n  class=\"ircc-cl-lib-footer footing\"\n  role=\"contentinfo\"\n>\n  <div\n    class=\"grid-container\"\n    [id]=\"id\"\n  >\n    <div class=\"item1 body3\">\n      <ng-content></ng-content>\n    </div>\n    <picture class=\"item2\">\n      <img\n        [src]=\"'data:image/png;base64,' + logo\"\n        alt=\"{{ altImage }}\"\n      />\n    </picture>\n  </div>\n</footer>\n", preserveWhitespaces: true });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.2.12", ngImport: i0, type: FooterComponent, decorators: [{
            type: Component,
            args: [{ selector: 'ircc-cl-lib-footer', template: "<footer\n  class=\"ircc-cl-lib-footer footing\"\n  role=\"contentinfo\"\n>\n  <div\n    class=\"grid-container\"\n    [id]=\"id\"\n  >\n    <div class=\"item1 body3\">\n      <ng-content></ng-content>\n    </div>\n    <picture class=\"item2\">\n      <img\n        [src]=\"'data:image/png;base64,' + logo\"\n        alt=\"{{ altImage }}\"\n      />\n    </picture>\n  </div>\n</footer>\n" }]
        }], ctorParameters: function () { return [{ type: i1.TranslateService }, { type: i2.LanguageHeaderFooterSwitchService }]; }, propDecorators: { id: [{
                type: Input
            }] } });
