// ircc-ds-angular-component-library/esm2020/lib/header-footer/language-switch/language-header-footer-switch.service.mjs
import { Injectable } from '@angular/core';
import { BehaviorSubject } from 'rxjs';
import * as i0 from "@angular/core";
/**
 * Language Switch Button renamed and refactored to include Header and Footer Dark Mode support based on user system preferences
 */
export class LanguageHeaderFooterSwitchService {
    constructor() {
        this.languageClickSub = new BehaviorSubject(false);
        this.languageClickObs$ = this.languageClickSub.asObservable();
        this.isDarkModeSubject = new BehaviorSubject(false);
        this.isDarkMode$ = this.isDarkModeSubject.asObservable();
        this.updateDarkModePreference();
    }
    /**
   * We check system preference for preferred color scheme in the constructor before components are rendered in the DOM and subscribe to changes in the header and footer components
   */
    updateDarkModePreference() {
        const darkModeQuery = window.matchMedia('(prefers-color-scheme: dark)');
        const darkModeListener = (event) => {
            this.isDarkModeSubject.next(event.matches);
        };
        darkModeQuery.addEventListener('change', darkModeListener);
        this.isDarkModeSubject.next(darkModeQuery.matches);
    }
    languageToggleClick() {
        this.languageClickSub.next(true);
    }
}
LanguageHeaderFooterSwitchService.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.2.12", ngImport: i0, type: LanguageHeaderFooterSwitchService, deps: [], target: i0.ɵɵFactoryTarget.Injectable });
LanguageHeaderFooterSwitchService.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "14.2.12", ngImport: i0, type: LanguageHeaderFooterSwitchService, providedIn: 'root' });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.2.12", ngImport: i0, type: LanguageHeaderFooterSwitchService, decorators: [{
            type: Injectable,
            args: [{
                    providedIn: 'root'
                }]
        }], ctorParameters: function () { return []; } });
