// ircc-ds-angular-component-library/esm2020/lib/shared/indicator/indicator.component.mjs
import { Component, Input, ViewChild } from '@angular/core';
import * as i0 from "@angular/core";
import * as i1 from "@angular/common";
import * as i2 from "../icon/icon.component";
import * as i3 from "@ngx-translate/core";
export var IndicatorType;
(function (IndicatorType) {
    IndicatorType[IndicatorType["dot"] = 0] = "dot";
    IndicatorType[IndicatorType["text"] = 1] = "text";
    IndicatorType[IndicatorType["number"] = 2] = "number";
})(IndicatorType || (IndicatorType = {}));
export var IndicatorTreatment;
(function (IndicatorTreatment) {
    IndicatorTreatment["strong"] = "strong";
    IndicatorTreatment["weak"] = "weak";
})(IndicatorTreatment || (IndicatorTreatment = {}));
export var IndicatorPurpose;
(function (IndicatorPurpose) {
    IndicatorPurpose["status"] = "status";
    IndicatorPurpose["palette"] = "palette";
})(IndicatorPurpose || (IndicatorPurpose = {}));
export var IndicatorStatus;
(function (IndicatorStatus) {
    IndicatorStatus["information"] = "information";
    IndicatorStatus["warning"] = "warning";
    IndicatorStatus["critical"] = "critical";
    IndicatorStatus["neutral"] = "neutral";
    IndicatorStatus["primary"] = "primary";
    IndicatorStatus["success"] = "success";
})(IndicatorStatus || (IndicatorStatus = {}));
export var IndicatorPalette;
(function (IndicatorPalette) {
    IndicatorPalette[IndicatorPalette["teal"] = 0] = "teal";
    IndicatorPalette[IndicatorPalette["orange"] = 1] = "orange";
    IndicatorPalette[IndicatorPalette["red"] = 2] = "red";
    IndicatorPalette[IndicatorPalette["grey"] = 3] = "grey";
    IndicatorPalette[IndicatorPalette["blue"] = 4] = "blue";
    IndicatorPalette[IndicatorPalette["green"] = 5] = "green";
    IndicatorPalette[IndicatorPalette["purple"] = 6] = "purple";
    IndicatorPalette[IndicatorPalette["navy"] = 7] = "navy";
})(IndicatorPalette || (IndicatorPalette = {}));
export class IndicatorComponent {
    constructor() {
        this.config = {
            type: 'text',
            category: IndicatorTreatment.weak,
            purpose: IndicatorPurpose.status,
            tabIndex: -1
        };
        this.tabIndex = undefined;
        this.EIndicatorStatus = IndicatorStatus;
    }
    ngOnInit() {
        //set config from individual options, if present
        if (this.size)
            this.config.size = this.size;
        if (this.type)
            this.config.type = this.type;
        if (this.icon)
            this.config.icon = this.icon;
        if (this.category)
            this.config.category = this.category;
        if (this.purpose)
            this.config.purpose = this.purpose;
        if (this.status)
            this.config.status = this.status;
        if (this.palette)
            this.config.palette = this.palette;
        if (this.ariaLabel)
            this.config.ariaLabel = this.ariaLabel;
        if (!this.tabIndex)
            this.config.tabIndex = -1;
        this.checkLabelRounded();
        this.checkNumber();
        this.tabIndex !== undefined
            ? (this.config.tabIndex = this.tabIndex)
            : undefined;
    }
    ngAfterViewInit() {
        setTimeout(() => this.checkLabelLength());
    }
    ngOnChanges(changes) {
        this.checkNumber();
        this.checkLabelRounded();
        this.checkLabelLength();
        setTimeout(() => this.checkLabelLength());
    }
    // Check if number exceeds 99
    checkNumber() {
        if (this.config.type === 'number' &&
            this.config?.label &&
            this.config.label > 99) {
            this.config.label = '99+';
        }
    }
    // If label only have 1 character, it should be rounded
    checkLabelRounded() {
        if (typeof this.config?.label === 'string') {
            this.rounded = this.config.label.length == 1 && !this.config.icon;
        }
    }
    // Check if div exceeds 200px
    checkLabelLength() {
        if (this.config.type !== 'text')
            return;
        // Max 200px - padding 8px x2
        this.abbr = ((this.label?.nativeElement?.offsetWidth &&
            this.label?.nativeElement?.offsetWidth > 184));
    }
}
IndicatorComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.2.12", ngImport: i0, type: IndicatorComponent, deps: [], target: i0.ɵɵFactoryTarget.Component });
IndicatorComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "14.2.12", type: IndicatorComponent, selector: "ircc-cl-lib-indicator", inputs: { config: "config", size: "size", type: "type", icon: "icon", category: "category", purpose: "purpose", status: "status", palette: "palette", ariaLabel: "ariaLabel", tabIndex: "tabIndex" }, viewQueries: [{ propertyName: "label", first: true, predicate: ["label"], descendants: true }], usesOnChanges: true, ngImport: i0, template: "<div\n  [ngClass]=\"['indicator-container', config.size ?? '']\"\n    [attr.aria-label]=\"\n    (config.ariaLabel ?? 'Indicator.Heading' | translate) +\n    ' ' +\n    (config.label ?? '')\n  \"\n  [tabindex]=\"config.tabIndex\"\n  role=\"group\"\n>\n  <div\n    *ngIf=\"config.type === 'text' || config.type === 'number'\"\n    #label\n    [ngClass]=\"[\n      config.type,\n      config.category,\n      config.purpose === 'status'\n        ? EIndicatorStatus[config.status ?? 'information']\n        : '',\n      config.purpose === 'palette' ? config.palette : '',\n      rounded ? 'rounded' : '',\n      config.label === '99+' ? 'num-lg' : '',\n      abbr ? 'abbr' : ''\n    ]\"\n  >\n    <ircc-cl-lib-icon\n      *ngIf=\"config.type === 'text' && config?.icon\"\n      [config]=\"{ FA_keywords: config.icon }\"\n      [attr.size]=\"config.size\"\n    ></ircc-cl-lib-icon>\n    <ng-container *ngIf=\"abbr; else nonAbbr\">\n      <span class=\"abbr\"\n        ><abbr [title]=\"config.label\">{{ config.label }}</abbr></span\n      >\n    </ng-container>\n    <ng-template #nonAbbr>\n      <span>{{ config.label }}</span>\n    </ng-template>\n  </div>\n  <span\n    *ngIf=\"config.type === 'dot'\"\n    [ngClass]=\"[\n      config.icon ? 'dot-icon' : '',\n      config.type,\n      config.category,\n      config.purpose === 'status'\n        ? EIndicatorStatus[config.status ?? 'information']\n        : '',\n      config.purpose === 'palette' ? config.palette : ''\n    ]\"\n  >\n    <ircc-cl-lib-icon\n      *ngIf=\"config?.icon\"\n      [config]=\"{ FA_keywords: config.icon }\"\n      [attr.size]=\"config.size\"\n    ></ircc-cl-lib-icon>\n  </span>\n</div>\n", dependencies: [{ kind: "directive", type: i1.NgClass, selector: "[ngClass]", inputs: ["class", "ngClass"] }, { kind: "directive", type: i1.NgIf, selector: "[ngIf]", inputs: ["ngIf", "ngIfThen", "ngIfElse"] }, { kind: "component", type: i2.IconComponent, selector: "ircc-cl-lib-icon", inputs: ["config", "ariaLabel", "FA_keywords", "size"] }, { kind: "pipe", type: i3.TranslatePipe, name: "translate" }], preserveWhitespaces: true });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.2.12", ngImport: i0, type: IndicatorComponent, decorators: [{
            type: Component,
            args: [{ selector: 'ircc-cl-lib-indicator', template: "<div\n  [ngClass]=\"['indicator-container', config.size ?? '']\"\n    [attr.aria-label]=\"\n    (config.ariaLabel ?? 'Indicator.Heading' | translate) +\n    ' ' +\n    (config.label ?? '')\n  \"\n  [tabindex]=\"config.tabIndex\"\n  role=\"group\"\n>\n  <div\n    *ngIf=\"config.type === 'text' || config.type === 'number'\"\n    #label\n    [ngClass]=\"[\n      config.type,\n      config.category,\n      config.purpose === 'status'\n        ? EIndicatorStatus[config.status ?? 'information']\n        : '',\n      config.purpose === 'palette' ? config.palette : '',\n      rounded ? 'rounded' : '',\n      config.label === '99+' ? 'num-lg' : '',\n      abbr ? 'abbr' : ''\n    ]\"\n  >\n    <ircc-cl-lib-icon\n      *ngIf=\"config.type === 'text' && config?.icon\"\n      [config]=\"{ FA_keywords: config.icon }\"\n      [attr.size]=\"config.size\"\n    ></ircc-cl-lib-icon>\n    <ng-container *ngIf=\"abbr; else nonAbbr\">\n      <span class=\"abbr\"\n        ><abbr [title]=\"config.label\">{{ config.label }}</abbr></span\n      >\n    </ng-container>\n    <ng-template #nonAbbr>\n      <span>{{ config.label }}</span>\n    </ng-template>\n  </div>\n  <span\n    *ngIf=\"config.type === 'dot'\"\n    [ngClass]=\"[\n      config.icon ? 'dot-icon' : '',\n      config.type,\n      config.category,\n      config.purpose === 'status'\n        ? EIndicatorStatus[config.status ?? 'information']\n        : '',\n      config.purpose === 'palette' ? config.palette : ''\n    ]\"\n  >\n    <ircc-cl-lib-icon\n      *ngIf=\"config?.icon\"\n      [config]=\"{ FA_keywords: config.icon }\"\n      [attr.size]=\"config.size\"\n    ></ircc-cl-lib-icon>\n  </span>\n</div>\n" }]
        }], propDecorators: { config: [{
                type: Input
            }], size: [{
                type: Input
            }], type: [{
                type: Input
            }], icon: [{
                type: Input
            }], category: [{
                type: Input
            }], purpose: [{
                type: Input
            }], status: [{
                type: Input
            }], palette: [{
                type: Input
            }], ariaLabel: [{
                type: Input
            }], tabIndex: [{
                type: Input
            }], label: [{
                type: ViewChild,
                args: ['label', { static: false }]
            }] } });
