// ircc-ds-angular-component-library/esm2020/lib/form-components/multi-checkbox/multi-checkbox.service.mjs
import { Injectable } from '@angular/core';
import { Subject } from 'rxjs';
import * as i0 from "@angular/core";
export class MultiCheckboxService {
    constructor() {
        //This is a subject, we're making subject => observable after
        // Anytime change to subject, it will propagate out to all other things related to it
        this.multiCheckboxEventSubj = new Subject();
        this.multiCheckboxEventObs$ = this.multiCheckboxEventSubj.asObservable(); //Use this for any events we need propagated up to parents
        this.multiCheckboxErrorSubj = new Subject();
        this.multiCheckboxErrorObs$ = this.multiCheckboxErrorSubj.asObservable();
    }
    /**
     * Broadcast element events
     * @param event: ImultiCheckboxEvent where id is the id of the component broadcasting and event is the Event
     */
    checkEvent(event) {
        this.multiCheckboxEventSubj.next(event);
    }
    errorEvent(event) {
        this.multiCheckboxErrorSubj.next(event);
    }
    /**
     * Checks to see if an AbstractControl is valid or not.
     * @param control: AbstractControl field that we are checking
     * @param field: string, name of the field we are checking
     */
    checkField(id) {
        //is the control valid
        this.errorEvent(id);
    }
}
MultiCheckboxService.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.2.12", ngImport: i0, type: MultiCheckboxService, deps: [], target: i0.ɵɵFactoryTarget.Injectable });
MultiCheckboxService.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "14.2.12", ngImport: i0, type: MultiCheckboxService, providedIn: 'root' });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.2.12", ngImport: i0, type: MultiCheckboxService, decorators: [{
            type: Injectable,
            args: [{
                    providedIn: 'root'
                }]
        }] });
