// ircc-ds-angular-component-library/esm2020/lib/form-components/multi-checkbox/multi-checkbox.component.mjs
import { Component, Input } from '@angular/core';
import { FormGroup } from '@angular/forms';
import * as i0 from "@angular/core";
import * as i1 from "../../../public-api";
import * as i2 from "../../../shared/functions/stand-alone.functions";
import * as i3 from "@angular/common";
import * as i4 from "../../shared/label/label.component";
import * as i5 from "../checkbox/checkbox.component";
import * as i6 from "../error/error.component";
import * as i7 from "@ngx-translate/core";
export class MultiCheckboxComponent {
    constructor(multiCheckboxService, standAloneFunctions) {
        this.multiCheckboxService = multiCheckboxService;
        this.standAloneFunctions = standAloneFunctions;
        this.form = new FormGroup({});
        this.errorIds = [];
        this.config = {
            id: '',
            label: {
                parentID: '',
                formGroup: this.form
            },
            errorMessages: []
        };
        this.id = '';
        /**
         * Accumulates all the errors on parents and children checkboxes
         * @type IErrorPairsMultiCheckBox[]
         */
        this.errorMessagesAccumulator = [];
        /**
         * Accumulates unique error types and used in the frontend to display
         * @type IErrorPairsMultiCheckBox[]
         */
        this.errorMessagesAccumulatorUniqueType = [];
        this.disabledStatus = false;
        this.groupCheckbox = true;
        if (this.id !== '')
            this.config.id = this.id;
        if (this.parent)
            this.config.parent = this.parent;
        if (this.children)
            this.config.children = this.children;
        if (this.errorMessages)
            this.config.errorMessages = this.errorMessages;
    }
    // returns size depending on if its group or multicheckbox
    get size() {
        if (this.groupCheckbox && this.config.children) {
            return this.config?.children[0]?.size || 'large';
        }
        return this.config.parent?.size || 'large';
    }
    ngOnInit() {
        //errorChecking
        this.checkErrorsSubscription();
        this.errorSub = this.multiCheckboxService.multiCheckboxErrorObs$.subscribe((response) => {
            if (this.config.children) {
                let children = this.config.children?.findIndex((child) => {
                    return child.id === response;
                });
                if (children > -1) {
                    this.checkError(this.config?.children[children]?.formGroup?.get(response)?.status, this.config?.children[children]?.formGroup || new FormGroup({}), response);
                }
            }
            if (this.config?.parent?.id === response) {
                this.checkError(this.config?.parent?.formGroup?.get(response)?.status, this.config?.parent?.formGroup || new FormGroup({}), response);
            }
        });
        if (this.config.parent !== undefined) {
            this.groupCheckbox = false;
            this.configSub =
                this.multiCheckboxService.multiCheckboxEventObs$.subscribe((response) => {
                    if (this.config.children) {
                        if (response.id === this.config.parent?.id) {
                            this.config.children?.forEach((res) => {
                                this.multiCheckboxService.checkEvent({
                                    id: res.id,
                                    event: response.event
                                });
                            });
                            this.config.parent.mixed = false;
                            this.checkError(this.config?.parent?.formGroup?.get(this.config.parent?.id)
                                ?.status, this.config?.parent?.formGroup || new FormGroup({}), this.config.parent?.id);
                            this.config.children?.forEach((res) => {
                                this.checkError(this.config?.parent?.formGroup?.get(res.id)?.status, this.config?.parent?.formGroup || new FormGroup({}), res.id);
                            });
                        }
                        else if (this.config.children?.findIndex((child) => {
                            return child.id === response.id;
                        }) > -1) {
                            let positive = 0;
                            let negative = 0;
                            this.config.children?.forEach((res) => {
                                res.formGroup.get(res.id)?.value === true
                                    ? positive++
                                    : negative++;
                            });
                            if (positive > 0 && negative > 0) {
                                this.config.parent?.formGroup
                                    .get(this.config.parent?.id)
                                    ?.patchValue(true, { emitEvent: false });
                                this.config.parent.mixed = true;
                            }
                            else if (positive > 0 && negative == 0) {
                                this.config.parent?.formGroup
                                    .get(this.config.parent?.id)
                                    ?.patchValue(true, { emitEvent: false });
                                this.config.parent.mixed = false;
                            }
                            else if (positive == 0 && negative > 0) {
                                this.config.parent?.formGroup
                                    .get(this.config.parent?.id)
                                    ?.patchValue(false, { emitEvent: false });
                                this.config.parent.mixed = false;
                            }
                            this.checkError(this.config?.parent?.formGroup?.get(this.config.parent?.id)
                                ?.status, this.config.parent?.formGroup || new FormGroup({}), this.config.parent?.id || '');
                            this.config.children?.forEach((res) => {
                                this.checkError(this.config?.parent?.formGroup?.get(res.id)?.status, this.config?.parent?.formGroup || new FormGroup({}), res.id);
                            });
                        }
                    }
                });
        }
    }
    checkErrorsSubscription() {
        this.config.children?.forEach((res) => {
            res.formGroup?.get(res.id)?.statusChanges.subscribe((value) => {
                this.checkError(value, res.formGroup, res.id);
                value === 'DISABLED'
                    ? (this.disabledStatus = true)
                    : (this.disabledStatus = false);
            });
        });
        if (this.config.parent !== undefined) {
            this.config.parent.formGroup
                ?.get(this.config.parent.id)
                ?.statusChanges.subscribe((value) => {
                this.checkError(value, this.config?.parent?.formGroup || new FormGroup({}), this.config?.parent?.id || '');
                value === 'DISABLED'
                    ? (this.disabledStatus = true)
                    : (this.disabledStatus = false);
                this.config.children?.forEach((res) => {
                    res.formGroup
                        ?.get(res.id)
                        ?.statusChanges.subscribe((value) => {
                        this.checkError(value, res.formGroup, res.id);
                        value === 'DISABLED'
                            ? (this.disabledStatus = true)
                            : (this.disabledStatus = false);
                    });
                });
            });
        }
    }
    /**
     * Will check the status of a formcontrol.  If the control is not in a valid status, it will loop through the errors
     * and add them to this.errorMessages.
     * @param value Status of the formcontrol
     * @param group - Formgroup to be checked
     * @param id - ID of the control that is being checked for error in the formgroup
     */
    checkError(value, group, id) {
        if (value !== 'VALID') {
            for (const error in group.get(id)?.errors) {
                let errorIndex = this.config.errorMessages?.findIndex((errorPair) => {
                    return errorPair.key === error;
                });
                let errorMessagesKeys = this.errorMessagesAccumulator?.findIndex((errorPair) => {
                    return errorPair.key === error && id === errorPair.id;
                });
                if (errorIndex > -1 && errorMessagesKeys === -1) {
                    this.errorMessagesAccumulator.push({
                        id: id,
                        key: this.config.errorMessages[errorIndex].key,
                        errorLOV: this.config.errorMessages[errorIndex].errorLOV
                    });
                }
                this.filterErrorList();
            }
        }
        else {
            this.errorMessagesAccumulator = this.errorMessagesAccumulator.filter((errorPair) => {
                return (errorPair.id)?.replace('_error0', '') !== id;
            });
            this.filterErrorList();
        }
    }
    /**
     * Will loop through the errors in the variable errorMessages.
     * It will then add a copy in errorMessagesAccumulator for each unique error type.
     * THe first error in errorMessagesAccumulator will get _error0 appended to it's ID to display the icon
     */
    filterErrorList() {
        this.errorMessagesAccumulatorUniqueType = [];
        this.errorMessagesAccumulator.forEach((error) => {
            let errorIndex = this.errorMessagesAccumulatorUniqueType.findIndex((errorPair) => {
                return errorPair.key === error.key;
            });
            if (errorIndex === -1) {
                this.errorMessagesAccumulatorUniqueType.push(error);
            }
        });
        if (this.errorMessagesAccumulatorUniqueType[0]) {
            if (this.errorMessagesAccumulatorUniqueType[0].id && !this.errorMessagesAccumulatorUniqueType[0].id.endsWith("_error0")) {
                this.errorMessagesAccumulatorUniqueType[0].id = this.errorMessagesAccumulatorUniqueType[0].id + "_error0";
            }
        }
    }
    isValid(checkbox) {
        return typeof checkbox !== 'undefined';
    }
    removeResponseFromError(response) {
        this.config.errorMessages = this.config.errorMessages.filter((item) => {
            item?.key !== response.id;
        });
    }
}
MultiCheckboxComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.2.12", ngImport: i0, type: MultiCheckboxComponent, deps: [{ token: i1.MultiCheckboxService }, { token: i2.StandAloneFunctions }], target: i0.ɵɵFactoryTarget.Component });
MultiCheckboxComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "14.2.12", type: MultiCheckboxComponent, selector: "ircc-cl-lib-multi-checkbox", inputs: { config: "config", id: "id", parent: "parent", children: "children", errorMessages: "errorMessages" }, ngImport: i0, template: "<fieldset>\n  <legend class=\"visually-hidden\">\n    {{ config.label.label || '' | translate }}\n  </legend>\n  <ng-container *ngIf=\"config.label\">\n    <ircc-cl-lib-label\n      [config]=\"config.label\"\n      [legend]=\"true\"\n      [attr.size]=\"size\"\n    ></ircc-cl-lib-label>\n  </ng-container>\n  <div class=\"parent-checkbox\">\n    <ng-container *ngIf=\"config.parent; else group\">\n      <ircc-cl-lib-checkbox [config]=\"config.parent\"></ircc-cl-lib-checkbox>\n      <ng-container *ngFor=\"let child of config.children\">\n        <ng-container\n          *ngTemplateOutlet=\"\n            recursiveListTmpl;\n            context: { checkBoxConfig: child }\n          \"\n        ></ng-container>\n      </ng-container>\n    </ng-container>\n  </div>\n\n  <ng-template #group>\n    <ng-container *ngFor=\"let child of config.children\">\n      <ircc-cl-lib-checkbox [config]=\"child\"></ircc-cl-lib-checkbox>\n    </ng-container>\n  </ng-template>\n\n  <ng-template\n    #recursiveListTmpl\n    let-checkBoxConfig=\"checkBoxConfig\"\n  >\n    <div class=\"child-checkbox\">\n      <ircc-cl-lib-checkbox [config]=\"checkBoxConfig\"></ircc-cl-lib-checkbox>\n    </div>\n  </ng-template>\n  <ng-container\n    *ngFor=\"let errors of errorMessagesAccumulatorUniqueType; let i = index\"\n  >\n    <div\n      *ngIf=\"errorMessagesAccumulatorUniqueType.length > -1 && !disabledStatus\"\n      class=\"radio-errors\"\n    >\n      <ircc-cl-lib-error\n        [id]=\"errors.id\"\n        [errorLOV]=\"errors.errorLOV\"\n        [attr.size]=\"size\"\n      ></ircc-cl-lib-error>\n    </div>\n  </ng-container>\n</fieldset>\n", styles: [".parent-checkbox{min-height:55px;margin-right:32px}.child-checkbox{margin-left:20px}.visually-hidden{position:absolute;width:1px;height:1px;padding:0;margin:-1px;overflow:hidden;clip:rect(0,0,0,0);white-space:nowrap;border:0}fieldset{border:none}\n"], dependencies: [{ kind: "directive", type: i3.NgForOf, selector: "[ngFor][ngForOf]", inputs: ["ngForOf", "ngForTrackBy", "ngForTemplate"] }, { kind: "directive", type: i3.NgIf, selector: "[ngIf]", inputs: ["ngIf", "ngIfThen", "ngIfElse"] }, { kind: "directive", type: i3.NgTemplateOutlet, selector: "[ngTemplateOutlet]", inputs: ["ngTemplateOutletContext", "ngTemplateOutlet", "ngTemplateOutletInjector"] }, { kind: "component", type: i4.LabelComponent, selector: "ircc-cl-lib-label", inputs: ["config", "formGroup", "errorMessages", "parentID", "label", "desc", "hint", "required", "iconButton", "topLabel", "touched", "legend"] }, { kind: "component", type: i5.CheckboxComponent, selector: "ircc-cl-lib-checkbox", inputs: ["config", "formGroup", "id", "label", "required", "size", "mixed", "inlineLabel", "inlineLabelBold", "hint", "desc", "errorMessages"] }, { kind: "component", type: i6.ErrorComponent, selector: "ircc-cl-lib-error", inputs: ["config", "id", "errorLOV", "icon", "size"] }, { kind: "pipe", type: i7.TranslatePipe, name: "translate" }], preserveWhitespaces: true });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.2.12", ngImport: i0, type: MultiCheckboxComponent, decorators: [{
            type: Component,
            args: [{ selector: 'ircc-cl-lib-multi-checkbox', template: "<fieldset>\n  <legend class=\"visually-hidden\">\n    {{ config.label.label || '' | translate }}\n  </legend>\n  <ng-container *ngIf=\"config.label\">\n    <ircc-cl-lib-label\n      [config]=\"config.label\"\n      [legend]=\"true\"\n      [attr.size]=\"size\"\n    ></ircc-cl-lib-label>\n  </ng-container>\n  <div class=\"parent-checkbox\">\n    <ng-container *ngIf=\"config.parent; else group\">\n      <ircc-cl-lib-checkbox [config]=\"config.parent\"></ircc-cl-lib-checkbox>\n      <ng-container *ngFor=\"let child of config.children\">\n        <ng-container\n          *ngTemplateOutlet=\"\n            recursiveListTmpl;\n            context: { checkBoxConfig: child }\n          \"\n        ></ng-container>\n      </ng-container>\n    </ng-container>\n  </div>\n\n  <ng-template #group>\n    <ng-container *ngFor=\"let child of config.children\">\n      <ircc-cl-lib-checkbox [config]=\"child\"></ircc-cl-lib-checkbox>\n    </ng-container>\n  </ng-template>\n\n  <ng-template\n    #recursiveListTmpl\n    let-checkBoxConfig=\"checkBoxConfig\"\n  >\n    <div class=\"child-checkbox\">\n      <ircc-cl-lib-checkbox [config]=\"checkBoxConfig\"></ircc-cl-lib-checkbox>\n    </div>\n  </ng-template>\n  <ng-container\n    *ngFor=\"let errors of errorMessagesAccumulatorUniqueType; let i = index\"\n  >\n    <div\n      *ngIf=\"errorMessagesAccumulatorUniqueType.length > -1 && !disabledStatus\"\n      class=\"radio-errors\"\n    >\n      <ircc-cl-lib-error\n        [id]=\"errors.id\"\n        [errorLOV]=\"errors.errorLOV\"\n        [attr.size]=\"size\"\n      ></ircc-cl-lib-error>\n    </div>\n  </ng-container>\n</fieldset>\n", styles: [".parent-checkbox{min-height:55px;margin-right:32px}.child-checkbox{margin-left:20px}.visually-hidden{position:absolute;width:1px;height:1px;padding:0;margin:-1px;overflow:hidden;clip:rect(0,0,0,0);white-space:nowrap;border:0}fieldset{border:none}\n"] }]
        }], ctorParameters: function () { return [{ type: i1.MultiCheckboxService }, { type: i2.StandAloneFunctions }]; }, propDecorators: { config: [{
                type: Input
            }], id: [{
                type: Input
            }], parent: [{
                type: Input
            }], children: [{
                type: Input
            }], errorMessages: [{
                type: Input
            }] } });
