// ircc-ds-angular-component-library/esm2020/lib/shared/dynamic-image/dynamic-image.component.mjs
import { Component, Input, ViewChild } from '@angular/core';
import { DSViewPortSize } from '../../../shared/constants/jl-components.constants';
import * as i0 from "@angular/core";
import * as i1 from "@ngx-translate/core";
export class DynamicImageComponent {
    constructor(translate, renderer, el) {
        this.translate = translate;
        this.renderer = renderer;
        this.el = el;
        this.config = {
            id: '',
            breakpoints: [{ maxWidth: DSViewPortSize.default, src: '' }],
            altText: '',
            defaultSrc: '',
            lazyLoad: false
        };
        this.id = '';
        this.altText = '';
        this.defaultSrc = '';
        this.lazyLoad = false;
    }
    ngOnInit() {
        if (this.id)
            this.config.id = this.id;
        if (this.breakpoints)
            this.config.breakpoints = this.breakpoints;
        if (this.altText)
            this.config.altText = this.altText;
        if (this.defaultSrc)
            this.config.defaultSrc = this.defaultSrc;
        if (this.lazyLoad)
            this.config.lazyLoad = this.lazyLoad;
    }
    ngAfterViewInit() {
        this.updateImageSrc();
        window.addEventListener('resize', () => this.updateImageSrc());
    }
    updateImageSrc() {
        if (this.image?.nativeElement) {
            const screenWidth = window.innerWidth;
            // Find the matching breakpoint for the current screen width
            const breakpoint = this.config.breakpoints.find((breakpoint) => screenWidth <= breakpoint.maxWidth);
            if (breakpoint) {
                // Set the src attribute of the image to the matched breakpoint's src
                this.renderer.setAttribute(this.image?.nativeElement, 'src', breakpoint.src);
            }
            else {
                // If no matching breakpoint is found, use the default src image
                this.renderer.setAttribute(this.image?.nativeElement, 'src', this.config.defaultSrc);
            }
        }
    }
}
DynamicImageComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.2.12", ngImport: i0, type: DynamicImageComponent, deps: [{ token: i1.TranslateService }, { token: i0.Renderer2 }, { token: i0.ElementRef }], target: i0.ɵɵFactoryTarget.Component });
DynamicImageComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "14.2.12", type: DynamicImageComponent, selector: "ircc-cl-lib-dynamic-image", inputs: { config: "config", id: "id", breakpoints: "breakpoints", altText: "altText", defaultSrc: "defaultSrc", lazyLoad: "lazyLoad" }, viewQueries: [{ propertyName: "image", first: true, predicate: ["image"], descendants: true, static: true }], ngImport: i0, template: "<div\n  id=\"{{ config.id + '-container' }}\"\n  class=\"dynamic-image-wrapper\"\n>\n  <img\n    #image\n    [alt]=\"config.altText || '' | translate\"\n    [attr.loading]=\"config.lazyLoad ? 'lazy' : null\"\n  />\n</div>\n", dependencies: [{ kind: "pipe", type: i1.TranslatePipe, name: "translate" }], preserveWhitespaces: true });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.2.12", ngImport: i0, type: DynamicImageComponent, decorators: [{
            type: Component,
            args: [{ selector: 'ircc-cl-lib-dynamic-image', template: "<div\n  id=\"{{ config.id + '-container' }}\"\n  class=\"dynamic-image-wrapper\"\n>\n  <img\n    #image\n    [alt]=\"config.altText || '' | translate\"\n    [attr.loading]=\"config.lazyLoad ? 'lazy' : null\"\n  />\n</div>\n" }]
        }], ctorParameters: function () { return [{ type: i1.TranslateService }, { type: i0.Renderer2 }, { type: i0.ElementRef }]; }, propDecorators: { image: [{
                type: ViewChild,
                args: ['image', { static: true }]
            }], config: [{
                type: Input
            }], id: [{
                type: Input
            }], breakpoints: [{
                type: Input
            }], altText: [{
                type: Input
            }], defaultSrc: [{
                type: Input
            }], lazyLoad: [{
                type: Input
            }] } });
