// ircc-ds-angular-component-library/esm2020/lib/header-footer/ircc-ds-angular-header-footer.module.mjs
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { HeaderComponent } from './header/header.component';
import { FooterComponent } from './footer/footer.component';
import { IrccDsAngularComponentsSharedModule } from '../shared/ircc-ds-angular-component-shared.module';
import { TranslateModule } from '@ngx-translate/core';
import { FormsModule, ReactiveFormsModule } from '@angular/forms';
import { HiddenNavComponent } from './hidden-nav/hidden-nav.component';
import { ThemeSwitchComponent } from './theme-switch/theme-switch.component';
import { LanguageSwitchComponent } from './language-switch/language-switch.component';
import * as i0 from "@angular/core";
const IrccDsHeaderFooterComponents = [
    HeaderComponent,
    FooterComponent,
    LanguageSwitchComponent,
    HiddenNavComponent,
    ThemeSwitchComponent
];
export class IrccDsAngularHeaderFooterModule {
}
IrccDsAngularHeaderFooterModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.2.12", ngImport: i0, type: IrccDsAngularHeaderFooterModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
IrccDsAngularHeaderFooterModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "14.2.12", ngImport: i0, type: IrccDsAngularHeaderFooterModule, declarations: [HeaderComponent,
        FooterComponent,
        LanguageSwitchComponent,
        HiddenNavComponent,
        ThemeSwitchComponent], imports: [CommonModule,
        IrccDsAngularComponentsSharedModule,
        TranslateModule,
        FormsModule,
        ReactiveFormsModule], exports: [HeaderComponent,
        FooterComponent,
        LanguageSwitchComponent,
        HiddenNavComponent,
        ThemeSwitchComponent] });
IrccDsAngularHeaderFooterModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "14.2.12", ngImport: i0, type: IrccDsAngularHeaderFooterModule, imports: [CommonModule,
        IrccDsAngularComponentsSharedModule,
        TranslateModule,
        FormsModule,
        ReactiveFormsModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.2.12", ngImport: i0, type: IrccDsAngularHeaderFooterModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [...IrccDsHeaderFooterComponents],
                    imports: [
                        CommonModule,
                        IrccDsAngularComponentsSharedModule,
                        TranslateModule,
                        FormsModule,
                        ReactiveFormsModule
                    ],
                    exports: [...IrccDsHeaderFooterComponents]
                }]
        }] });
